lib/portlayer/exec/base.go

```diff
@@ -231,7 +233,37 @@ func (c *containerBase) stop(ctx context.Context, waitTime *int32) error {
 
 	log.Warnf("stopping %s via hard power off due to: %s", c.ExecConfig.ID, err)
 
-	return c.poweroff(ctx)
+	err = c.poweroff(ctx)
+
+	// In case of InvalidPowerState and GenericVmConfigFault during hard poweroff, nil would be returned.
+	// At the time the VM must've been already powered off. But we may need to wait for extra time
+	// such that we can obtain the latest powerstate from the VC.
+	if err == nil {
```

Comment:
Should we retry in the InvalidPowerState case here? Also in the `GenericVmConfigFault` case we should probably return the error from `poweroff()`, so that we can do this extra wait here. Or do the actual wait in `poweroff()`?

Reply:
For the case of `GenericVmConfigFault`, in `poweroff()`, if the error msg indicates `vmNotSuspendedKey` or `vmPoweringOffKey`, `poweroff()` returns `nil` and the retry logic will be executed. Otherwise, the error will be returned and the retry logic will not be called. So I think it should be fine to either have the retry logic here instead or inside `poweroff()`.

Reply:
`poweroff()` returns nil in this case: https://sourcegraph.com/github.com/chengwang86/vic@ab324c9076cbf71e94fac1532e6d5e7a5b945ea9/-/blob/lib/portlayer/exec/base.go#L373-374 . We should not retry in this case.